File: c619/b.py
t = int(input())
for i in range(t):
	n = int(input())
	l = [int(x) for x in input().split()]
	ne = 0
	ll = []
	for j in range(n):
		if l[j] != -1:
			ll.append(l[j])
			ne+=1
	if ne == 0:
		a = 1
	else:
		a = int((max(ll)+min(ll))/2)
	for j in range(n):
		if l[j] == -1:
			l[j] = a
	m = 0
	for j in range(1,n):
		if abs(l[j]-l[j-1])>m:
			m = abs(l[j]-l[j-1])
	print(m,a)

File: c619/b.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long
#define pb push_back
int main(int argc, char const *argv[])
{
	int t;
	cin>>t;
	for (int i = 0; i < t; ++i)
		{
			int n;
			cin>>n;
			ll a[n];
			ll nn = 0;
			ll s = 0;
			ll mx = 0;
			ll mn = 1e9+5;
			for (int i = 0; i < n; ++i)
			{
				cin>>a[i];
				if (a[i] != -1)
				{
					mx = max(mx,a[i]);
					mn = min(mn,a[i]);
				}
			}
			ll k = 0;
			if (nn == 0)
			{
				k = 1;
			}
			else{
				k = s/nn;
			}
			ll m = 0;
			if (a[0] == -1)
			{
				a[0] = k;
			}
			for (int i = 1; i < n; ++i)
			{
				if (a[i] == -1)
				{
					a[i] = k;
				}
				m = max(m,abs(a[i]-a[i-1]));
			}
			cout<<m<<" "<<k<<endl;
		}	
	return 0;
}